core/services/keeper/orm.go

```diff
@@ -99,25 +109,45 @@ DELETE FROM upkeep_registrations WHERE registry_id IN (
 	return rowsAffected, nil
 }
 
-func (korm ORM) EligibleUpkeepsForRegistry(registryAddress ethkey.EIP55Address, blockNumber, gracePeriod int64) (upkeeps []UpkeepRegistration, err error) {
+//EligibleUpkeepsForRegistry fetches eligible upkeeps for processing
+//The query checks the following conditions
+// - checks the registry address is correct and the registry has some keepers associated
+// -- is it my turn AND my keeper was not the last perform for this upkeep OR my keeper was the last before BUT it is past the grace period
+// -- OR is it my buddy's turn AND they were the last keeper to do the perform for this upkeep
+func (korm ORM) EligibleUpkeepsForRegistry(registryAddress ethkey.EIP55Address, head *types.Head, gracePeriod int64) (upkeeps []UpkeepRegistration, err error) {
+	registry, err := korm.RegistryByContractAddress(registryAddress)
+	if err != nil {
+		return nil, errors.Wrap(err, "EligibleUpkeepsForRegistry failed to get a registry by address")
+	}
+	blockNumber := head.Number
+	binaryHash := binaryOfFirstHashInTurn(blockNumber, registry, head)
+
 	stmt := `
 SELECT upkeep_registrations.* FROM upkeep_registrations
 INNER JOIN keeper_registries ON keeper_registries.id = upkeep_registrations.registry_id
 WHERE
 	keeper_registries.contract_address = $1 AND
-	keeper_registries.num_keepers > 0 AND
-	(
-		upkeep_registrations.last_run_block_height = 0 OR (
-			upkeep_registrations.last_run_block_height + $2 < $3 AND
-			upkeep_registrations.last_run_block_height < ($3 - ($3 % keeper_registries.block_count_per_turn))
-		)
-	) AND
-	keeper_registries.keeper_index = (
-		upkeep_registrations.positioning_constant + (($3 - ($3 % keeper_registries.block_count_per_turn)) / keeper_registries.block_count_per_turn)
-	) % keeper_registries.num_keepers
-ORDER BY upkeep_registrations.id ASC, upkeep_registrations.upkeep_id ASC
+	keeper_registries.num_keepers > 0 AND 
+    ((
+                keeper_registries.keeper_index = ((CAST(upkeep_registrations.upkeep_id AS bit(32)) #
+                                                   CAST($4 AS bit(32)))::bigint % keeper_registries.num_keepers)
+            AND
+                (
+				upkeep_registrations.last_keeper_index IS DISTINCT FROM keeper_registries.keeper_index
+				OR
+				(upkeep_registrations.last_keeper_index IS NOT DISTINCT FROM keeper_registries.keeper_index AND upkeep_registrations.last_run_block_height + $2 < $3)
```

Comment:
curious, what's the difference from using '=' instead of IS NOT DISTINCT

Reply:
we need to use it when doing comparison against NULL. 

> Ordinary comparison operators yield null (signifying “unknown”), not true or false, when either input is null. For example, 7 = NULL yields null, as does 7 <> NULL. When this behavior is not suitable, use the IS [ NOT ] DISTINCT FROM predicates.

> Do not write expression = NULL because NULL is not “equal to” NULL.

from https://www.postgresql.org/docs/current/functions-comparison.html